feat: support audio language selector for deepgram API, add fallback for both openai/deepgram

jfouret opened the pull request:
# Pull Request Checklist

### Note to first-time contributors: Please open a discussion post in [Discussions](https://github.com/open-webui/open-webui/discussions) and describe your changes before submitting a pull request.

**Before submitting, make sure you've checked the following:**

- [x] **Target branch:** Please verify that the pull request targets the `dev` branch.
- [x] **Description:** Provide a concise description of the changes made in this pull request.
- [x] **Changelog:** Ensure a changelog entry following the format of [Keep a Changelog](https://keepachangelog.com/) is added at the bottom of the PR description.
- [x] **Documentation:** Have you updated relevant documentation [Open WebUI Docs](https://github.com/open-webui/docs), or other documentation sources?
- [x] **Dependencies:** Are there any new dependencies? Have you updated the dependency versions in the documentation?
- [x] **Testing:** Have you written and run sufficient tests to validate the changes?
- [x] **Code review:** Have you performed a self-review of your code, addressing any coding standard issues and ensuring adherence to the project's coding standards?
- [x] **Prefix:** To clearly categorize this pull request, prefix the pull request title using one of the following:
  - **BREAKING CHANGE**: Significant changes that may affect compatibility
  - **build**: Changes that affect the build system or external dependencies
  - **ci**: Changes to our continuous integration processes or workflows
  - **chore**: Refactor, cleanup, or other non-functional code changes
  - **docs**: Documentation update or addition
  - **feat**: Introduces a new feature or enhancement to the codebase
  - **fix**: Bug fix or error correction
  - **i18n**: Internationalization or localization changes
  - **perf**: Performance improvement
  - **refactor**: Code restructuring for better maintainability, readability, or scalability
  - **style**: Changes that do not affect the meaning of the code (white space, formatting, missing semi-colons, etc.)
  - **test**: Adding missing tests or correcting existing tests
  - **WIP**: Work in progress, a temporary label for incomplete or ongoing work

# Changelog Entry

### Description

The language selector support mentioned in #13989 was not properly implemented for deepgram API.

### Added

- Support audio language parameter for deepgram API
- Add fallback if language parameter is not recognized for openai and deepgram api

---

### Additional Information

- Implement a fallback mecanismfor failed API calls linked with bad language
- Handle 400 errors by removing language param and retrying with next language candidate
- Add warning logs for language parameter removal attempts

here the logs from my tests (test done on v0.6.18 within docker)
I juste updated the audio.py file in the backend repository
openAI:
```
2025-07-26 13:51:31.709 | INFO     | open_webui.routers.audio:transcription:948 - file.content_type: audio/webm; codecs=opus - {}
2025-07-26 13:51:31.710 | INFO     | open_webui.routers.audio:transcribe:827 - transcribe: /app/backend/data/cache/audio/transcriptions/e0615551-4a01-4c1f-9643-c5bc0bd68367.webm {'language': 'iougiug'} - {}
2025-07-26 13:51:32.139 | INFO     | open_webui.routers.audio:convert_audio_to_mp3:116 - Converted /app/backend/data/cache/audio/transcriptions/e0615551-4a01-4c1f-9643-c5bc0bd68367.webm to /app/backend/data/cache/audio/transcriptions/e0615551-4a01-4c1f-9643-c5bc0bd68367.mp3 - {}
Chunk paths: ['/app/backend/data/cache/audio/transcriptions/e0615551-4a01-4c1f-9643-c5bc0bd68367.mp3']
2025-07-26 13:51:33.599 | WARNING  | open_webui.routers.audio:transcription_handler:602 - Failed openai transcribe with language="iougiug", Retrying with "None"... - {}
2025-07-26 13:51:34.537 | INFO     | uvicorn.protocols.http.httptools_impl:send:476 - 176.144.33.197:50860:0 - "POST /api/v1/audio/transcriptions HTTP/1.1" 200 - {}
2025-07-26 13:51:38.169 | INFO     | uvicorn.protocols.http.httptools_impl:send:476 - 176.144.33.197:50860:0 - "GET /_app/version.json HTTP/1.1" 304 - {}
2025-07-26 13:51:42.156 | INFO     | uvicorn.protocols.http.httptools_impl:send:476 - 176.144.33.197:50860:0 - "GET /_app/version.json HTTP/1.1" 304 - {}
2025-07-26 13:51:44.852 | INFO     | uvicorn.protocols.http.httptools_impl:send:476 - 176.144.33.197:50860:0 - "GET /api/v1/chats/99ffa00f-d9b0-4df3-9182-38b6d6a0df2a HTTP/1.1" 200 - {}
2025-07-26 13:51:44.871 | INFO     | uvicorn.protocols.http.httptools_impl:send:476 - 176.144.33.197:50860:0 - "GET /api/v1/chats/99ffa00f-d9b0-4df3-9182-38b6d6a0df2a HTTP/1.1" 200 - {}
2025-07-26 13:51:44.904 | INFO     | uvicorn.protocols.http.httptools_impl:send:476 - 176.144.33.197:50860:0 - "GET /api/v1/chats/99ffa00f-d9b0-4df3-9182-38b6d6a0df2a HTTP/1.1" 200 - {}
2025-07-26 13:52:09.633 | INFO     | uvicorn.protocols.http.httptools_impl:send:476 - 176.144.33.197:50860:0 - "GET /_app/version.json HTTP/1.1" 304 - {}
```

Deepgram:
```
2025-07-26 13:57:55.715 | INFO     | uvicorn.protocols.http.httptools_impl:send:476 - 176.144.33.197:50860:0 - "GET /api/v1/chats/all/tags HTTP/1.1" 200 - {}
2025-07-26 13:57:59.178 | INFO     | open_webui.routers.audio:transcription:949 - file.content_type: audio/webm; codecs=opus - {}
2025-07-26 13:57:59.179 | INFO     | open_webui.routers.audio:transcribe:828 - transcribe: /app/backend/data/cache/audio/transcriptions/9cff6b64-e751-4899-be3b-53b3052a0cb2.webm {'language': 'iougiug'} - {}
2025-07-26 13:57:59.607 | INFO     | open_webui.routers.audio:convert_audio_to_mp3:116 - Converted /app/backend/data/cache/audio/transcriptions/9cff6b64-e751-4899-be3b-53b3052a0cb2.webm to /app/backend/data/cache/audio/transcriptions/9cff6b64-e751-4899-be3b-53b3052a0cb2.mp3 - {}
Chunk paths: ['/app/backend/data/cache/audio/transcriptions/9cff6b64-e751-4899-be3b-53b3052a0cb2.mp3']
2025-07-26 13:57:59.995 | WARNING  | open_webui.routers.audio:transcription_handler:668 - Failed deepgram transcribe with language="iougiug", Retrying with "None"... - {}
2025-07-26 13:58:00.876 | INFO     | uvicorn.protocols.http.httptools_impl:send:476 - 176.144.33.197:50860:0 - "POST /api/v1/audio/transcriptions HTTP/1.1" 200 - {}
2025-07-26 13:58:08.468 | INFO     | uvicorn.protocols.http.httptools_impl:send:476 - 176.144.33.197:50860:0 - "GET /api/usage HTTP/1.1" 200 - {}
2025-07-26 13:58:10.111 | INFO     | uvicorn.protocols.http.httptools_impl:send:476 - 176.144.33.197:50860:0 - "GET /_app/version.json HTTP/1.1" 200 - {}
2025-07-26 13:58:12.413 | INFO     | uvicorn.protocols.http.httptools_impl:send:476 - 176.144.33.197:50860:0 - "GET /api/v1/audio/voices HTTP/1.1" 200 - {}
2025-07-26 13:58:18.467 | INFO     | open_webui.routers.openai:get_all_models:392 - get_all_models() - {}
2025-07-26 13:58:18.804 | INFO     | uvicorn.protocols.http.httptools_impl:send:476 - 176.144.33.197:50860:0 - "GET /api/models HTTP/1.1" 200 - {}
2025-07-26 13:58:18.996 | INFO     | uvicorn.protocols.http.httptools_impl:send:476 - 176.144.33.197:50860:0 - "POST /api/v1/users/user/settings/update HTTP/1.1" 200 - {}
2025-07-26 13:58:30.767 | INFO     | uvicorn.protocols.http.httptools_impl:send:476 - 176.144.33.197:50860:0 - "GET /api/v1/chats/6a01aad5-8e89-4753-8a12-b2d74c6358c0 HTTP/1.1" 200 - {}
2025-07-26 13:58:31.099 | INFO     | uvicorn.protocols.http.httptools_impl:send:476 - 176.144.33.197:50860:0 - "GET /api/v1/chats/9cb8b7ee-e459-42cf-8bd8-704231e8caf1 HTTP/1.1" 200 - {}
2025-07-26 13:58:31.100 | INFO     | uvicorn.protocols.http.httptools_impl:send:476 - 176.144.33.197:50860:0 - "GET /api/v1/chats/0181d97a-5ac5-4d02-833a-b0e4fd02f596 HTTP/1.1" 200 - {}
2025-07-26 13:58:32.855 | INFO     | open_webui.routers.audio:transcription:949 - file.content_type: audio/webm; codecs=opus - {}
2025-07-26 13:58:32.856 | INFO     | open_webui.routers.audio:transcribe:828 - transcribe: /app/backend/data/cache/audio/transcriptions/9a17270d-5e2f-4b00-b9e0-29f822abc4d8.webm {'language': 'multi'} - {}
2025-07-26 13:58:33.434 | INFO     | open_webui.routers.audio:convert_audio_to_mp3:116 - Converted /app/backend/data/cache/audio/transcriptions/9a17270d-5e2f-4b00-b9e0-29f822abc4d8.webm to /app/backend/data/cache/audio/transcriptions/9a17270d-5e2f-4b00-b9e0-29f822abc4d8.mp3 - {}
Chunk paths: ['/app/backend/data/cache/audio/transcriptions/9a17270d-5e2f-4b00-b9e0-29f822abc4d8.mp3']
2025-07-26 13:58:34.675 | INFO     | uvicorn.protocols.http.httptools_impl:send:476 - 176.144.33.197:50860:0 - "POST /api/v1/audio/transcriptions HTTP/1.1" 200 - {}
```

I tried with French and it works, below after transcription with deepgram nova-3 (multi).

<img width="1092" height="146" alt="image" src="https://github.com/user-attachments/assets/84f20187-86dc-4407-8829-90f848c031ae" />

This match the  #13720 which was closed too fast.

Note that I did not find any test suite related to `audio.py`
Note that local whisper does not raise any error when a randome string is given as language parameters, so I did not implement any fallback here. Azure logic seems different as we can submit several loavles already.
I tried to integrate comment from @rgaricano about #15935 

### Contributor License Agreement

By submitting this pull request, I confirm that I have read and fully agree to the [Contributor License Agreement (CLA)](/CONTRIBUTOR_LICENSE_AGREEMENT), and I am providing my contributions under its terms.

jfouret committed:
Add retry logic and language parameter handling for Deepgram API

- Add language parameter from metadata to Deepgram requests
- Implement 3-retry mechanism for failed API calls
- Handle 400 errors by removing language param and retrying
- Add warning logs for language parameter removal attempts

rgaricano commented:
I suggest a language choice similar to the one made in PR (https://github.com/open-webui/open-webui/pull/15935), which forces the preferred language to be the language environment variable if it exists, even if the audio metadata indicates a different one.

jfouret commented:
> I suggest a language choice similar to the one made in PR (#15935), which forces the preferred language to be the language environment variable if it exists, even if the audio metadata indicates a different one.

The variable introduced in this PR is used here. Did you mean something else ?

jfouret committed:
Refactor language selection logic in transcription handler

Extract repeated language selection logic into a single variable to reduce code duplication across different API calls.

jfouret commented:
> I suggest a language choice similar to the one made in PR (#15935), which forces the preferred language to be the language environment variable if it exists, even if the audio metadata indicates a different one.

I added your code at the begining to reuse in all engines.

jfouret commented:
Will retest later with this modif.

rgaricano commented on backend/open_webui/routers/audio.py:
```diff
@@ -594,8 +595,8 @@ def transcription_handler(request, file_path, metadata):
-                        {"language": metadata.get("language")}
-                        if metadata.get("language")
+                        {"language": selected_language}
+                        if selected_language
```
not necessary, it's implicit in selected_language

rgaricano commented on backend/open_webui/routers/audio.py:
```diff
@@ -647,12 +648,28 @@ def transcription_handler(request, file_path, metadata):
+                        "Retrying deepgram API call without the language param:"
+                        f" \"{params['language']}\""
+                    )
+                    del params["language"]
```
why?

jfouret replied:
imagine you setup `en-GB` with whisper and then you move to deepgram, As this is no supported and language param, it will return error 400 (nova-3 support only `en`/`en-us` or `multi`). It will make the API work even if a wrong languge param is selected

jfouret replied:
I limited the retry mecanism to case where langue is mentioned in the body with status code 400.

jfouret committed:
Implement language fallback for STT transcription requests

Replace single language selection with prioritized language list that falls back through metadata language, configured language, and auto-detection when transcription fails due to unsupported language errors in OpenAI and Deepgram STT engines.

jfouret commented:
I have proposed a other way of doing it, and did some tests. @rgaricano ?

rgaricano commented:
I think that you're overcomplicating it,
just the logic: selected_language=...

And there language it is assigned: if it is defined in whisper var language, that one is used,
if it is not in whisper var, then the one from the file's metadata is used if it exists, if it doesn't exist, None is assigned.
And in the rest of the functions, just use selected_language, regardless of its value, if it is None the function will use the default option (auto) and if not, it's going to use the one assigned in the previous logic.

jfouret commented:
Sorry, could you specify what exactly do I overcomplicate ?
- the `languages` list in place of selected_language ?
- the fallback mecanism when a language parameter is not accepted by the model/provider to avoid runtime errors ?

rgaricano commented:
> the fallback mecanism when a language parameter is not accepted by the model/provider to avoid runtime errors ?

I don't see any fallback, just the same openai & deepgram that before, and its works fine with that param.
Really all STT/TTS engines have a standard language param (maybe with differences in admited locale codes & except azure that use locales and regions).

jfouret commented:
As an admin we know, It was more for a user perspective.

I have introduced this fallback for the following case:
- As admin I change from openai to deepgram
- As user I have se my language to french (FR) in my personal settings, which is accepted by openai, but not deepgram which expects `language="multi"` for non-english support for the model `nova-3`.
- No fallback results then in error 400
- Fallback to None is good
- Falling to multi set by admin (env WHISPER_LANGUAGE) is better if the admin knows main language preference on the instance.

rgaricano commented:
Yes, but whisper doesn't support "multi", maybe it's better is the fallback have some logit for that, e.g. in deepgram check for locales that have to match with multi and then replace by multi.
( my setting is ES, & I'm using default whisper)

rgaricano commented:
thanks, I was refering to global one, there isn't that option in adminSettings

![open-webui-whisper0](https://github.com/user-attachments/assets/642bab72-5058-4963-99e5-e7ab58a3b22c)

jfouret commented:
> Closing in favour of [f23eb2a](https://github.com/open-webui/open-webui/commit/f23eb2a31c3874f5ed14b81385fd4296206dbc84)

After removing my local patch and moving to the .26 version, it's no longer working.

The request received by deepgram `/v1/listen?smart_format=true&model=nova-3`

I can fix it **and** test it if you want @tjbck 

Bests
